ui/src/app/shared/url-search-params.utils.ts

```diff
@@ -0,0 +1,103 @@
+import {URLSearchParams} from '@angular/http';
+
+import {QueryJobsRequest} from "./model/QueryJobsRequest";
+import {JobStatus} from "./model/JobStatus";
+import {QueryFields} from "./common";
+
+/** Utilities for working with URLSearchParams*/
+export class URLSearchParamsUtils {
+
+  /** Accepts a QueryJobsRequest, and translates it reversibly into a string that can be passed as a QueryParam. */
+  public static encodeURLSearchParams(request: QueryJobsRequest): string {
+    let urlSearchParams = new URLSearchParams();
+    urlSearchParams.set(QueryFields.parentId, request.parentId);
+    urlSearchParams.set(QueryFields.jobName, request.name);
+    for (let s in request.statuses) {
+      urlSearchParams.append('statuses', JobStatus[request.statuses[s]]);
+    }
+    if (request.start) {
+      urlSearchParams.set(QueryFields.start, request.start.toISOString());
+    }
+    if (request.end) {
+      urlSearchParams.set(QueryFields.end, request.end.toISOString());
+    }
+
+    if (request.labels) {
+      (request.labels as Map<String, String>).forEach((value: string, key: string) => {
+        urlSearchParams.set(key, value);
+      });
+    }
+    return urlSearchParams.toString();
+  }
+
+  /** Accepts a param map, and translates it reversibly into a string that can be passed as a QueryParam. */
+  public static encodeURLSearchParamsFromMap(params: Map<String, String[]>): string {
+    let urlSearchParams = new URLSearchParams();
+    params.forEach((values: string[], key: string) => {
+      if (values.length == 1) {
+        urlSearchParams.set(key, values[0]);
+      } else {
+        values.forEach(value => urlSearchParams.append(key, value));
+      }
+    });
+    return urlSearchParams.toString();
+  }
+
+  /** Accepts a string query that was generated via encodeUrlSearchParams and converts it back into a QueryJobsRequest. */
+  public static unpackURLSearchParams(query: string): QueryJobsRequest {
+    let urlSearchParams = new URLSearchParams(query);
+    let queryRequest = {
+      labels: new Map(),
```

Comment:
Why did you chose a `Map` over key value object?

Reply:
See other comment